include/nbl/video/IGPUDescriptorSet.h

```diff
@@ -112,49 +118,39 @@ class NBL_API IGPUDescriptorSet : public asset::IDescriptorSet<const IGPUDescrip
-                    continue;
+        inline core::smart_refctd_ptr<asset::IDescriptor>* getAllDescriptors(const asset::IDescriptor::E_TYPE type) const
+        {
+            auto* baseAddress = m_pool->getDescriptorStorage(type);
+            if (baseAddress == nullptr)
+                return nullptr;
 
-                const auto type = static_cast<asset::E_DESCRIPTOR_TYPE>(i);
-                std::destroy_n(getDescriptorStorage(type) + m_descriptorStorageOffsets.data[i], m_layout->getTotalDescriptorCount(type));
-            }
+            const auto offset = m_descriptorStorageOffsets.data[static_cast<uint32_t>(type)];
+            if (offset == ~0u)
+                return nullptr;
 
-            const auto mutableSamplerCount = m_layout->getTotalMutableSamplerCount();
-            if (mutableSamplerCount > 0)
-                std::destroy_n(getMutableSamplerStorage() + getMutableSamplerStorageOffset(), mutableSamplerCount);
-		}
+            return baseAddress + offset;
+        }
 
-	private:
-        struct SDescriptorOffsets
+        inline core::smart_refctd_ptr<IGPUSampler>* getAllMutableSamplers() const
         {
-            SDescriptorOffsets()
-            {
-                // The default constructor should initiailze all the offsets to an invalid value (~0u) because ~IGPUDescriptorSet relies on it to
-                // know which descriptors are present in the set and hence should be destroyed.
-                std::fill_n(data, asset::EDT_COUNT + 1, ~0u);
-            }
+            auto* baseAddress = m_pool->getMutableSamplerStorage();
+            if (baseAddress == nullptr)
+                return nullptr;
+
+            const auto poolOffset = m_descriptorStorageOffsets.data[static_cast<uint32_t>(asset::IDescriptor::E_TYPE::ET_COUNT)];
+            if (poolOffset == ~0u)
+                return nullptr;
 
-            uint32_t data[asset::EDT_COUNT + 1];
-        };
+            return baseAddress + poolOffset;
+        }
 
-        // Returns the offset into the pool's descriptor storage. These offsets will be combined
-        // later with base memory addresses to get the actual memory address where we put the core::smart_refctd_ptr<const IDescriptor>.
-        void allocateDescriptors() override;
-        core::smart_refctd_ptr<asset::IDescriptor>* getDescriptorStorage(const asset::E_DESCRIPTOR_TYPE type) const;
-        core::smart_refctd_ptr<IGPUSampler>* getMutableSamplerStorage() const;
+        inline uint32_t getDescriptorStorageOffset(const asset::IDescriptor::E_TYPE type) const { return m_descriptorStorageOffsets.data[static_cast<uint32_t>(type)]; }
+        inline uint32_t getMutableSamplerStorageOffset() const { return m_descriptorStorageOffsets.data[static_cast<uint32_t>(asset::IDescriptor::E_TYPE::ET_COUNT)]; }
 
         std::atomic_uint64_t m_version;
-		core::smart_refctd_ptr<IDescriptorPool> m_pool;
-		SDescriptorOffsets m_descriptorStorageOffsets;
+        const uint32_t m_parentPoolVersion;
```

Comment:
If we implement https://github.com/Devsh-Graphics-Programming/Nabla/pull/345/files/8fbc87079edc5109a9c916eeb7ea7c0eff9ce128..b3719ddb1377d8af8f8d5f4dd112d3c844394553#r1116359755 then instead of `m_parentPoolVersion` we could keep an `m_isZombie` and let `IDescriptorSet` set that via friendship.

Reply:
actually you neither need `m_parentPoolVersion` nor `m_isZombie` if the pools keep weak pointers to the allocated descriptor sets

you can simply check `!m_pool` for zombie-ism (because there's no point for a zombie to hold onto its parent pool and keep it alive)

Reply:
Resolved.